### PR TITLE
e2e: add contexts and e2e generator tests

### PR DESCRIPTION
**Pull request by cmwaters**
This PR tackles some of the preliminary work in regards to the e2e outlined in  https://github.com/tendermint/tendermint/issues/9396

### CODE REVIEW

File: test/e2e/README.md [hunk 1]
```diff
@@ -7,7 +7,27 @@ make
 ./build/runner -f networks/ci.toml
 ```
 
-This creates and runs a testnet named `ci` under `networks/ci/` (determined by the manifest filename).
+This creates and runs a testnet named `ci` under `networks/ci/`.
+
+## Conceptual Overview
+
+End-to-end testnets are used to test Tendermint functionality as a user would use it, by spinning up a set of nodes with various configurations and making sure the nodes and network behave correctly. The background for the E2E test suite is outlined in [RFC-001](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-066-e2e-testing.md).
```
Comment (sergio-mena) [inline test/e2e/README.md:14]: RFC001 seems to refer to storage, not e2e. Am I missing something?

Comment (cmwaters) [reply test/e2e/README.md:14]: Yeah I guess this was missed in the earlier PR

**Comment by cmwaters** [comment]
This PR isn't based on cherry-picked commits. There were probably a handful of PRs. I've just directly looked at the diff between v0.36 and main and copied the relevant items across. This seems to be adding contexts to e2e methods, a test for the generator, more documentation and addition of extra data in the benchmark command

**Comment by sergio-mena** [comment]
> This PR isn't based on cherry-picked commits. There were probably a handful of PRs. I've just directly looked at the diff between v0.36 and main and copied the relevant items across. This seems to be adding contexts to e2e methods, a test for the generator, more documentation and addition of extra data in the benchmark command

What was the exact "git diff" you ran? (IOW, which two hashes were you comparing?)

**Comment by cmwaters** [comment]
> What was the exact "git diff" you ran? (IOW, which two hashes were you comparing?)

main and v0.36.x. (do you want the exact hashes?)

**Comment by sergio-mena** [comment]
> do you want the exact hashes?

_A priori_, not necessary, I'll "git diff" using the tip of both branches. If I run into ambiguities, I'll let you know.
Hopefully, I'll be able to tackle this next week.

**Review by thanethomson — COMMENTED** [review]
It seems like all this change does is plumb contexts through the E2E tests?

If so, the issue title and description don't really match up with what's being done here. Are there other changes from v0.36 that will eventually be ported across?

**Comment by sergio-mena** [comment]
Sorry for taking some to time come back to this.
I compared the diff between `v0.36.x` and `main`, with this PR. There seem to be large parts that are not present in this PR. I can understand that the vote extension related ones don't make sense (yet), but about about the others?

File: test/e2e/runner/benchmark.go [hunk 2]
```diff
@@ -19,45 +21,58 @@ import (
 //
 // Metrics are based of the `benchmarkLength`, the amount of consecutive blocks
 // sampled from in the testnet
-func Benchmark(testnet *e2e.Testnet, benchmarkLength int64) error {
-	block, _, err := waitForHeight(testnet, 0)
+func Benchmark(ctx context.Context, testnet *e2e.Testnet, benchmarkLength int64) error {
+	block, _, err := waitForHeight(ctx, testnet, 0)
```
[the remark this answers is not in the document]
Comment (cmwaters) [reply test/e2e/runner/benchmark.go:25]: No this just simply waits until all nodes have started before beginning the actual benchmark

**Comment by cmwaters** [comment]
> It seems like all this change does is plumb contexts through the E2E tests?

That's the crux of it. There were some more documenting changes; a test was added to ensure that all generated testnets are valid; modifications to how we wait for nodes to reach a certain height was adjusted (a lot of small stuff)

> If so, the issue title and description don't really match up with what's being done here. Are there other changes from v0.36 that will eventually be ported across?

Yeah, there are still a chunk of changes that can only happen once finalize block and vote extensions have actually been implemented. This PR is supposed to address the "preliminary work" within e2e as outlined by Sergio. There will definitely need to be a second PR later on. I'm fine with renaming it to something more scoped.

File: test/e2e/node/config.go [hunk 3]
```diff
@@ -23,7 +23,6 @@ type Config struct {
 	PrivValServer    string                      `toml:"privval_server"`
 	PrivValKey       string                      `toml:"privval_key"`
 	PrivValState     string                      `toml:"privval_state"`
-	Misbehaviors     map[string]string           `toml:"misbehaviors"`
```
Comment (thanethomson) [inline test/e2e/node/config.go:26]: So this doesn't currently support misbehavior? (Doesn't look like this was used anywhere)

**Review by thanethomson — APPROVED** [review]

File: test/e2e/node/config.go (hunk 3 above)
Comment (cmwaters) [reply test/e2e/node/config.go:26]: Yeah I think this was vestige from v0.34 back when the maverick node was around. It's unused.

**Pull request merged**